### PR TITLE
Read pdf hash feature

### PR DESCRIPTION
**Pull request by anas-yousef**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/XSUP-32377

## Description
Added support for extracting hashes.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by anas-yousef** [comment]
@JasBeilin We just need to decide how we want to update the documentation of the script.

File: Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py [hunk 1]
```diff
@@ -27,7 +26,6 @@ class PdfPermissionsException(Exception):
     Every exception class that is in charge of catching errors that occur when trying to
     extract data from the PDF must inherit this class
     """
-    pass
```
Comment (JasBeilin) [inline Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py:30]: Why removing it?

File: Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py [hunk 2]
```diff
@@ -36,15 +34,13 @@ class PdfCopyingProtectedException(PdfPermissionsException):
     a `copy-protected` file (Copy-protected files are files that prevent us from copy its content)
     This is relevant since we run a command that copies the content of the pdf file into a text file.
     """
-    pass
```
Comment (JasBeilin) [inline Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py:39]: ?

File: Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py [hunk 3]
```diff
@@ -36,15 +34,13 @@ class PdfCopyingProtectedException(PdfPermissionsException):
...
 
 
 class PdfInvalidCredentialsException(PdfPermissionsException):
     """
     This class is in charge of catching errors that occur when we try to decrypt an encrypted
     pdf file with the wrong password.
     """
-    pass
```
Comment (JasBeilin) [inline Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py:47]: ?

File: Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py [hunk 4]
```diff
@@ -545,6 +548,41 @@ def extract_urls_and_emails_from_pdf_file(file_path: str, output_folder: str) ->
...
+
+def convert_hash_to_context(hash_type: str, hashes: set[Any]) -> list[dict[str, Any]]:
+    hash_context: list[dict[str, Any]] = []
+    for hash in hashes:
+        hash_context.append({'type': hash_type,
+                             'value': hash})
+    return hash_context
+
+
+def get_hashes_from_file(file_path: str) -> dict[str, set[Any]]:
```
Comment (JasBeilin) [inline Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py:569]: Please add docstring

File: Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py [hunk 5]
```diff
@@ -545,6 +548,41 @@ def extract_urls_and_emails_from_pdf_file(file_path: str, output_folder: str) ->
...
+    demisto.debug('Extracting hashes from file')
+    hashes: dict[str, set[Any]] = {}
+    with open(file_path, 'rb') as pdf_file:
+        pdf = PyPDF2.PdfReader(pdf_file, strict=False)
```
Comment (JasBeilin) [inline Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py:573]: Why opening again the pdffile? Why not run on the text extracted from it?

File: Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py [hunk 6]
```diff
@@ -545,6 +548,41 @@ def extract_urls_and_emails_from_pdf_file(file_path: str, output_folder: str) ->
...
+            hash_contexts.extend(convert_hash_to_context(hash_type, hashes))
+    return hash_contexts
+
+
+def convert_hash_to_context(hash_type: str, hashes: set[Any]) -> list[dict[str, Any]]:
+    hash_context: list[dict[str, Any]] = []
+    for hash in hashes:
+        hash_context.append({'type': hash_type,
+                             'value': hash})
+    return hash_context
```
Comment (JasBeilin) [inline Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py:566]:
```suggestion
def convert_hash_to_context(hash_type: str, hashes: set[Any]) -> list[dict[str, Any]]:
    hash_context = [{'type': hash_type, 'value': hash} for hash in hashes]
    return hash_context
```

File: Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py [hunk 7]
```diff
@@ -545,6 +548,41 @@ def extract_urls_and_emails_from_pdf_file(file_path: str, output_folder: str) ->
     return urls_ec, emails_ec
 
 
+def extract_hash_contexts_from_pdf_file(file_path: str) -> list[dict[str, Any]]:
+    hash_contexts: list[dict[str, Any]] = []
+    # Get hashes from the binary file
+    hashes_in_file = get_hashes_from_file(file_path)
+    for hash_type, hashes in hashes_in_file.items():
+        if hashes:
+            hash_contexts.extend(convert_hash_to_context(hash_type, hashes))
```
Comment (JasBeilin) [inline Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py:557]: You can check it before the for loop and then have it a one liner, but its not critical.

File: Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py [hunk 8]
```diff
@@ -572,6 +610,8 @@ def extract_data_from_pdf(path: str, user_password: str, entry_id: str, max_imag
                                                  dec_file_path=f'{working_dir}/DecWorkingReadPDF.pdf',
                                                  encrypted=encrypted,
                                                  user_password=user_password)
+        # Get hash contexts
+        hash_contexts = extract_hash_contexts_from_pdf_file(cpy_file_path)
```
Comment (JasBeilin) [inline Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py:614]: Why running here and not on the text file like the URL and EMAIL? Lets keep the code clean and consistent.

File: Packs/CommonScripts/Scripts/ReadPDFFileV2/test_data/pdf-with-hashes.pdf
Comment (JasBeilin) [inline Packs/CommonScripts/Scripts/ReadPDFFileV2/test_data/pdf-with-hashes.pdf:1]: Lets make sure our example has both hashes and ip/url something else to make sure they are unaffected/ not detected as hash by mistake.

**Review by JasBeilin — CHANGES_REQUESTED** [review]
Nice work, lets enhance a bit and merge :)

File: Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py (hunk 1 above)
Comment (anas-yousef) [reply Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py:30]: Ruff

File: Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py (hunk 2 above)
Comment (anas-yousef) [reply Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py:39]: [Ruff](https://docs.astral.sh/ruff/rules/unnecessary-placeholder/)

File: Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py (hunk 3 above)
Comment (anas-yousef) [reply Packs/CommonScripts/Scripts/ReadPDFFileV2/ReadPDFFileV2.py:47]: [Ruff](https://docs.astral.sh/ruff/rules/unnecessary-placeholder/)

**Review by JasBeilin — APPROVED** [review]

**Pull request merged**